### PR TITLE
Stop password reset for unaccepted invitations

### PR DESCRIPTION
**Pull request by benthorner**
https://trello.com/c/HXSZikzB/38-improve-signon

Previously a user could request a password reset against an invitation that had not been accepted, even if it has expired. This ensures we only send password reset links to users with active accounts (no invitation, or invitation accepted).

### CODE REVIEW

File: app/views/user_mailer/notify_reset_password_disallowed_due_to_unaccepted_invitation.html.erb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<p>Your <%= link_to t('department.name'), t('department.url') %> Signon <%= account_name %>, for <%= @user.email %>, is has not been activated. You can't request a password reset on an inactive account.</p>
```
Comment (tijmenb) [inline app/views/user_mailer/notify_reset_password_disallowed_due_to_unaccepted_invitation.html.erb]:
```suggestion
<p>Your <%= link_to t('department.name'), t('department.url') %> Signon <%= account_name %>, for <%= @user.email %>, has not been activated. You can't request a password reset on an inactive account.</p>
```

File: app/views/user_mailer/notify_reset_password_disallowed_due_to_unaccepted_invitation.text.erb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+Your <%= link_to t('department.name'), t('department.url') %> Signon <%= account_name %>, for <%= @user.email %>, is has not been activated. You can't request a password reset on an inactive account.
```
Comment (tijmenb) [inline app/views/user_mailer/notify_reset_password_disallowed_due_to_unaccepted_invitation.text.erb]:
```suggestion
Your <%= link_to t('department.name'), t('department.url') %> Signon <%= account_name %>, for <%= @user.email %>, has not been activated. You can't request a password reset on an inactive account.
```

File: test/integration/password_reset_test.rb [hunk 3]
```diff
@@ -27,6 +27,30 @@ class PasswordResetTest < ActionDispatch::IntegrationTest
...
 
+  should "not allow password reset for unaccepted invitations" do
+    perform_enqueued_jobs do
+      user = create(:user, invitation_token: SecureRandom.uuid, invitation_accepted_at: nil)
+      trigger_reset_for(user.email)
+      assert_response_contains(BLANKET_RESET_MESSAGE)
+
+      open_email(user.email)
+      assert current_email
+      assert_equal "Your GOV.UK Signon development invitation has not been accepted", current_email.subject
```
Comment (tijmenb) [inline test/integration/password_reset_test.rb]:
```suggestion
      assert_equal "Your GOV.UK Signon development account has not been activated", current_email.subject
```

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -43,6 +43,11 @@ def notify_reset_password_disallowed_due_to_suspension(user)
     mail(to: @user.email, subject: suspension_notification_subject)
   end
 
+  def notify_reset_password_disallowed_due_to_unaccepted_invitation(user)
+    @user = user
+    mail(to: @user.email, subject: "Your #{app_name} invitation has not been accepted")
```
Comment (tijmenb) [inline app/mailers/user_mailer.rb]:
```suggestion
    mail(to: @user.email, subject: "Your #{app_name} account has not been activated")
```

**Comment by benthorner** [comment]
Thanks for spotting - should be fixed now :-).

**Pull request merged**